File: core/modules/categories/category.service.ts
import type { Category } from '../../../types/index.js';
import type { ICategoryRepository } from './category.repository.interface.js';

let categoryRepository: ICategoryRepository;

export function initCategoryService(repo: ICategoryRepository) {
  categoryRepository = repo;
}

/**
 * Build a tree structure from flat categories
 * @param {Array} categories - Flat category list
 * @param {number|null} parentId - Parent ID to filter by
 * @returns {Array} Tree structure
 */
function buildCategoryTree(categories: Category[], parentId: number | null = null): Category[] {
    const tree: Category[] = [];
    categories
        .filter(category => category.parent_id === parentId)
        .forEach(category => {
            const children = buildCategoryTree(categories, category.id);
            if (children.length > 0) {
                category.subcategories = children;
            }
            tree.push(category);
        });
    return tree;
}

/**
 * Get categories as a tree structure
 * @returns {Array} Category tree
 */
export function getCategoriesAsTree() {
    const allCategories = categoryRepository.findAllCategories();
    return buildCategoryTree(allCategories);
}

/**
 * Get all categories (flat)
 * @returns {Array} All categories
 */
export function getAllCategories() {
    return categoryRepository.findAllCategories();
}

/**
 * Get a category by ID
 * @param {number} id - Category ID
 * @returns {Object|null} Category or null
 */
export function getCategoryById(id: number) {
    return categoryRepository.findCategoryById(id);
}

/**
 * Create a new category
 * @param {string} name - Category name
 * @param {number|null} parentId - Parent category ID
 * @returns {Object} Created category
 */
export function createCategory(name: string, parentId: number | null = null) {
    return categoryRepository.insertCategory(name, parentId);
}

/**
 * Update a category
 * @param {number} id - Category ID
 * @param {Object} data - Update data
 * @returns {Object} Update result
 */
export function updateCategory(id: number, data: { name?: string; parentId?: number | null }) {
    return categoryRepository.updateCategoryById(id, data);
}

/**
 * Delete a category
 * @param {number} id - Category ID
 * @returns {Object} Delete result
 */
export function deleteCategory(id: number) {
    return categoryRepository.softDeleteCategory(id);
}

File: src/main/java/com/possum/application/categories/CategoryService.java
package com.possum.application.categories;

import com.possum.domain.exceptions.NotFoundException;
import com.possum.domain.exceptions.ValidationException;
import com.possum.domain.model.Category;
import com.possum.domain.repositories.CategoryRepository;

import java.util.ArrayList;
import java.util.List;

public class CategoryService {
    private final CategoryRepository categoryRepository;

    public CategoryService(CategoryRepository categoryRepository) {
        this.categoryRepository = categoryRepository;
    }

    public List<CategoryTreeNode> getCategoriesAsTree() {
        List<Category> allCategories = categoryRepository.findAllCategories();
        return buildCategoryTree(allCategories, null);
    }

    public List<Category> getAllCategories() {
        return categoryRepository.findAllCategories();
    }

    public Category getCategoryById(long id) {
        return categoryRepository.findCategoryById(id)
                .orElseThrow(() -> new NotFoundException("Category not found"));
    }

    public Category createCategory(String name, Long parentId) {
        com.possum.application.auth.ServiceSecurity.requirePermission(com.possum.application.auth.Permissions.CATEGORIES_MANAGE);
        if (name == null || name.trim().isEmpty()) {
            throw new ValidationException("Category name is required");
        }
        return categoryRepository.insertCategory(name, parentId);
    }

    public void updateCategory(long id, String name, Long parentId) {
        com.possum.application.auth.ServiceSecurity.requirePermission(com.possum.application.auth.Permissions.CATEGORIES_MANAGE);
        if (name == null || name.trim().isEmpty()) {
            throw new ValidationException("Category name is required");
        }
        boolean parentIdProvided = parentId != null;
        int changes = categoryRepository.updateCategoryById(id, name, parentIdProvided, parentId);
        if (changes == 0) {
            throw new NotFoundException("Category not found");
        }
    }

    public void deleteCategory(long id) {
        com.possum.application.auth.ServiceSecurity.requirePermission(com.possum.application.auth.Permissions.CATEGORIES_MANAGE);
        int changes = categoryRepository.softDeleteCategory(id);
        if (changes == 0) {
            throw new NotFoundException("Category not found");
        }
    }

    private List<CategoryTreeNode> buildCategoryTree(List<Category> categories, Long parentId) {
        List<CategoryTreeNode> tree = new ArrayList<>();
        categories.stream()
                .filter(category -> {
                    if (parentId == null) {
                        return category.parentId() == null;
                    }
                    return parentId.equals(category.parentId());
                })
                .forEach(category -> {
                    List<CategoryTreeNode> children = buildCategoryTree(categories, category.id());
                    tree.add(new CategoryTreeNode(category, children));
                });
        return tree;
    }

    public record CategoryTreeNode(Category category, List<CategoryTreeNode> subcategories) {
    }
}
